### DISCUSSION TITLE
Fixing Windows port

### DISCUSSION THREAD

**Discussion by asalwa** [ccache/ccache#817] [Development]
Hello. I'm trying to make Ccache work (again) on Windows.
I've created a draft pull request https://github.com/ccache/ccache/pull/816.
I'd like Ccache maintainer to look at these changes and discuss next steps.

In pull request I described errors which I've already found and fixed. 
All these errors are around execution of compiler on Windows, especially when handling long command line.

Some fixes are simple, and some require discussion before proceeding with pull.

I had to change declarations of functions execute() and win32execute(), because I needed to pass a path to temporary directory. I decided to add `Context&` as a parameter, as it contains both temp dir path and `pid_t` which is necessary on Linux.
But the ugly part of this change is that win32execute is called directly in one place, in cache_compilation() - at the end there is "if (fall_back_to_original_compiler)" where execv() is used. And execv() is a macro for Windows.
And this is why I added saved_temp_dir variable in cache_compilation().
Suggestion welcome on how can I make this code cleaner.

Regarding change in Win32Util::argv_to_string:
I've checked it with Mingw (mingw-w64/i686-8.1.0-posix-dwarf-rt_v6-rev0). When backslashes are present in args file, for example when calling a preprocessor, let's assume content of args file:
`"-E" "C:\Users\asalwa\test.cpp"`
And then command:
`g++ @cmd_args.ju6vOI`
Fails:
`g++.exe: error: C:Usersasalwatest.cpp: No such file or directory`
When I convert to double backslashes in args file it works fine (it turns out that gcc is processing backslashes in this file).

But behavior of Win32Util::argv_to_string should not be changed unconditionally. This change should be smarter. Maybe it should be parametrized, something like:
`std::string argv_to_string(const char* const* argv, const std::string& prefix, bool escape_backslashes = false);`

And then test case in unittest/test_Win32Util.cpp shall be extended to cover this behavior.

BTW: one unit test is failing on Windows:

>===============================================================================
C:/Users/asalwa/ccache-from-git/unittest/test_Util.cpp:564:
TEST SUITE: Util
TEST CASE:  Util::make_relative_path
  Path matches neither actual nor apparent CWD
>
>C:/Users/asalwa/ccache-from-git/unittest/test_Util.cpp:593: ERROR: CHECK( make_relative_path("C:/", "C:/a", "C:/b", "C:/x") == "C:/x" ) is NOT correct!
  values: CHECK( ..x == C:/x )
>
>===============================================================================
[doctest] test cases:  151 |  150 passed | 1 failed | 0 skipped

Please share your thoughts. I'd like to proceed with these changes, clean it as necessary to get it merged to master.

**Comment by asalwa** [comment]
Should I create an issue and link it to milestone 4.2.1 ?

**Reply by jrosdahl** [reply]
You can create a bug report if you like, but I'm fine without it as well. If you do you can't link it to milestone 4.2.1, though – that can only for be done by users with developer status. It wouldn't make sense for other users to be able to assign milestones since those are used for planning purposes.

**Comment by jrosdahl** [comment]
> Hello. I'm trying to make Ccache work (again) on Windows.

Thanks. 

> I had to change declarations of functions execute() and win32execute(), because I needed to pass a path to temporary directory. I decided to add `Context&` as a parameter, as it contains both temp dir path and `pid_t` which is necessary on Linux.

That's OK.

> But the ugly part of this change is that win32execute is called directly in one place, in cache_compilation() - at the end there is "if (fall_back_to_original_compiler)" where execv() is used. And execv() is a macro for Windows. And this is why I added saved_temp_dir variable in cache_compilation(). Suggestion welcome on how can I make this code cleaner.

I agree that's ugly. :slightly_smiling_face: I'll write a suggestion in a pull request review comment.

> [...] But behavior of Win32Util::argv_to_string should not be changed unconditionally. This change should be smarter. Maybe it should be parametrized, something like:
> `std::string argv_to_string(const char* const* argv, const std::string& prefix, bool escape_backslashes = false);`
> 
> And then test case in unittest/test_Win32Util.cpp shall be extended to cover this behavior.

Sounds OK.

> BTW: one unit test is failing on Windows: [...]

FWIW, it works for me in Wine.

**Reply by asalwa** [reply]
I've updated my pull request, please take a look.
